--- a/quickassist/utilities/osal/src/linux/kernel_space/OsalCryptoInterface.c	2020-09-09 15:31:18.000000000 +0200
+++ b/quickassist/utilities/osal/src/linux/kernel_space/OsalCryptoInterface.c	2020-10-07 08:18:54.224835720 +0200
@@ -66,7 +66,7 @@
 
 #include "Osal.h"
 #include <linux/crypto.h>
-#include <linux/cryptohash.h>
+#include <crypto/sha.h>
 #include <linux/version.h>
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,29))
 #include <crypto/internal/hash.h>
