--- a/quickassist/qat/drivers/crypto/qat/qat_common/qat_crypto.h	2020-09-09 15:30:08.000000000 +0200
+++ b/quickassist/qat/drivers/crypto/qat/qat_common/qat_crypto.h	2020-10-08 19:12:06.216300954 +0200
@@ -80,11 +80,11 @@
 	struct icp_qat_fw_la_bulk_req req;
 	union {
 		struct qat_alg_aead_ctx *aead_ctx;
-		struct qat_alg_ablkcipher_ctx *ablkcipher_ctx;
+		struct qat_alg_skcipher_ctx *skcipher_ctx;
 	};
 	union {
 		struct aead_request *aead_req;
-		struct ablkcipher_request *ablkcipher_req;
+		struct skcipher_request *skcipher_req;
 	};
 	struct qat_crypto_request_buffs buf;
 	void (*cb)(struct icp_qat_fw_la_resp *resp,
